sbom/spdx/encoder.go

```diff
@@ -0,0 +1,444 @@
+	creators := make([]v2common.Creator, len(e.Creators))
+	for i, creator := range e.Creators {
+		creators[i] = v2common.Creator{
+			Creator:     creator.Creator,
+			CreatorType: creator.CreatorType,
+		}
+	}
+
+	// Initial metadata
+	out := &v2_3.Document{
+		SPDXVersion:       v2_3.Version,
+		DataLicense:       v2_3.DataLicense,
+		SPDXIdentifier:    "DOCUMENT",
+		DocumentName:      e.DocumentName,
+		DocumentNamespace: e.DocumentNamespace,
+		CreationInfo: &v2_3.CreationInfo{
+			Creators: creators,
+			Created:  time.Now().Format("2006-01-02T15:04:05Z"),
+		},
+		DocumentComment: e.DocumentComment,
+	}
+
+	pkgs := make(map[int]*v2_3.Package)
+	var pkgIds []int
+	dists := make(map[int]*v2_3.Package)
+	var distIds []int
+	repos := make(map[int]*v2_3.Package)
+	var repoIds []int
+	pkgRels := map[int][]*v2_3.Relationship{}
+	for _, r := range ir.IndexRecords() {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		rPkgId, err := strconv.Atoi(r.Package.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		pkg, ok := pkgs[rPkgId]
+		// Record the package if we haven't seen it yet.
+		if !ok {
+			pkgDB := ""
+			for _, env := range ir.Environments[r.Package.ID] {
+				if env.PackageDB != "" {
+					pkgDB = env.PackageDB
+					break
+				}
+			}
+
+			pkgPurpose := "APPLICATION"
+			if r.Package.Kind != claircore.BINARY {
+				pkgPurpose = "SOURCE"
+			}
+
+			pkg = newSpdxPackageFromPackage(r.Package)
+			pkg.PackageFileName = pkgDB
+			pkg.FilesAnalyzed = true
+			pkg.PrimaryPackagePurpose = pkgPurpose
+
+			pkgs[rPkgId] = pkg
+			pkgIds = append(pkgIds, rPkgId)
+
+			if r.Package.Source != nil && r.Package.Source.Name != "" {
+				rSrcPkgId, err := strconv.Atoi(r.Package.Source.ID)
+				if err != nil {
+					return nil, err
+				}
+
+				srcPkg, ok := pkgs[rSrcPkgId]
+				// Record the source package if we haven't seen it yet.
+				if !ok {
+					srcPkg = newSpdxPackageFromPackage(r.Package.Source)
+					srcPkg.PrimaryPackagePurpose = "SOURCE"
+					pkgs[rSrcPkgId] = srcPkg
+					pkgIds = append(pkgIds, rSrcPkgId)
+				}
+
+				rel := &v2_3.Relationship{
+					RefA:         v2common.MakeDocElementID("", string(pkg.PackageSPDXIdentifier)),
+					RefB:         v2common.MakeDocElementID("", string(srcPkg.PackageSPDXIdentifier)),
+					Relationship: "GENERATED_FROM",
+				}
+				pkgRels[rPkgId] = append(pkgRels[rPkgId], rel)
+			}
+		} else if pkg.PrimaryPackagePurpose == "SOURCE" {
+			// If we recorded a source package when we found it as an r.Package.Source,
+			// we need to record any missing information we didn't know about previously.
+			pkg.FilesAnalyzed = true
+			if pkg.PackageFileName == "" {
+				pkgDB := ""
+				for _, env := range ir.Environments[r.Package.ID] {
+					if env.PackageDB != "" {
+						pkgDB = env.PackageDB
+						break
+					}
+				}
+				pkg.PackageFileName = pkgDB
+			}
+		}
+
+		// Record Distributions for this package.
+		if r.Distribution != nil {
+			rDistId, err := strconv.Atoi(r.Distribution.ID)
```

Comment:
just realized, do we use the IDs for anything other than converting them into `int`s? If not, then can we just keep them as strings and make the map key string instead of int? Same idea for the slices. Not sure if we have to keep doing this conversion

Reply:
We convert them to `int`s to sort them. The return from `IndexReport.IndexRecords()` is non-deterministic because it iterates over `IndexReport.Packages`, which is a map. In order to have a deterministic output, both to test against and for users to be able to run a diff between reports, we need to have some sort of ordering structuring. This is what I came up with, but I'm happy to revisit how we do it (perhaps in a follow up PR 🙂)

Some of this context is capture in a comment here: Some context here: https://github.com/quay/claircore/pull/1468/files#diff-9730b2b5baa2c38c549f0b3c1077fddf654896f316b79c2c8f3f77d5fb7ccd1cR273-R276

Reply:
so if we just cared about sorting for determinism, then I think we can keep it as strings. If we also want to have the IDs in increasing order, then converting to int makes sense. Feel free to change in a followup if you decide to change it